services/src/main/java/org/keycloak/userprofile/config/DeclarativeUserProfileProvider.java

```diff
@@ -305,14 +320,18 @@ protected UserProfileMetadata decorateUserProfileForCache(UserProfileMetadata me
             } else {
                 // always add validation for imuttable/read-only attributes
                 validators.add(new AttributeValidatorMetadata(ImmutableAttributeValidator.ID));
-                decoratedMetadata.addAttribute(attributeName, validators, writeAllowed, required, readAllowed).addAnnotations(annotations);
+                decoratedMetadata.addAttribute(attributeName, validators, selector, writeAllowed, required, readAllowed).addAnnotations(annotations);
             }
         }
 
         return decoratedMetadata;
 
     }
 
+    private boolean isCoreAttribute(String attributeName) {
```

Comment:
We already have `org.keycloak.userprofile.Attributes#isRootAttribute`. Looks like here we are interested only in username and email.

I would rename the method to `isUsernameOrEmailAttribute` or something.

Reply:
Makes sense, updated